### PR TITLE
Enable the EBS CSI driver

### PR DESCRIPTION
**Pull request by arjunrn**
This PR adds [CSI components](https://kubernetes-csi.github.io/docs/) for Kubernetes with the [EBS driver](https://github.com/kubernetes-sigs/aws-ebs-csi-driver).
 Also changed in this PR is the `prometheus` statefulset. Another statefulset needs to be created because the `persistentVolumeClaims` cannot be modified. Instead a new `statefulset` is created with CSI volumes.

### CODE REVIEW

**Comment by hjacobs** [comment]
Can you link to more information on what this is about?

File: cluster/cluster.yaml [hunk 1]
```diff
@@ -1340,8 +1340,61 @@ Resources:
...
+                - - 'arn:aws:iam::{{.Cluster.InfrastructureAccount | getAWSAccountID}}:role/'
+                  - !Ref MasterIAMRole
+          - Action:
+              - 'sts:AssumeRole'
+            Effect: Allow
+            Principal:
+              AWS: !Join
+                - ''
+                - - 'arn:aws:iam::{{.Cluster.InfrastructureAccount | getAWSAccountID}}:role/'
+                  - !Ref WorkerIAMRole
```
Comment (aermakov-zalando) [inline cluster/cluster.yaml:1369]: Do we really need it? We'll run the controller on the masters, so the workers shouldn't be able to delete volumes.

File: cluster/manifests/03-ebs-csi/02-csi-driver.yaml [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            periodSeconds: 10
+            timeoutSeconds: 3
+          name: ebs-plugin
+          ports:
+            - containerPort: 9808
+              name: healthz
+              protocol: TCP
+          resources:
+            requests:
+              cpu: 50m
```
Comment (aermakov-zalando) [inline cluster/manifests/03-ebs-csi/02-csi-driver.yaml]: Does it really need `50m`? This might be out of our budget in some of the clusters, and most of the time it shouldn't even do anything.

File: cluster/manifests/03-ebs-csi/controller-aws-iam-role.yaml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+{{ if eq .ConfigItems.kube_aws_iam_controller_kube_system_enable "true"}}
```
Comment (aermakov-zalando) [inline cluster/manifests/03-ebs-csi/controller-aws-iam-role.yaml]: We haven't had any issues with it, so I'd drop the config item. We can do this in a separate PR, but I wouldn't bother with it for new things.

File: cluster/manifests/03-ebs-csi/ebs-controller.yaml [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+      containers:
+        - args:
+            - controller
+            - --endpoint=$(CSI_ENDPOINT)
+            - --logtostderr
+            - --v=5
+          env:
+            - name: CSI_ENDPOINT
+              value: unix:///var/lib/csi/sockets/pluginproxy/csi.sock
+          {{ if eq .ConfigItems.kube_aws_iam_controller_kube_system_enable "true"}}
```
Comment (aermakov-zalando) [inline cluster/manifests/03-ebs-csi/ebs-controller.yaml]: Same here. Let's just use one mechanism.

File: cluster/manifests/03-ebs-csi/02-csi-driver.yaml [hunk 5]
```diff
@@ -0,0 +1,127 @@
+apiVersion: apps/v1
```
Comment (aermakov-zalando) [inline cluster/manifests/03-ebs-csi/02-csi-driver.yaml:2]: Since we're just testing it, let's make it conditional? Deploying it to all clusters might not be feasible because we might not have enough space on the nodes.

File: cluster/manifests/03-ebs-csi/02-csi-driver.yaml [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+  template:
+    metadata:
+      labels:
+        application: container-storage-interface
+        component: driver
+    spec:
+      containers:
+        - args:
+            - node
+            - --endpoint=$(CSI_ENDPOINT)
```
Comment (aermakov-zalando) [inline cluster/manifests/03-ebs-csi/02-csi-driver.yaml:43]: Didn't know you could do that!

File: cluster/manifests/03-ebs-csi/02-csi-driver.yaml [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+            limits:
+              cpu: 50m
+              memory: 50Mi
+          securityContext:
+            privileged: true
+          volumeMounts:
+            - mountPath: /opt/podruntime/kubelet
+              mountPropagation: Bidirectional
+              name: kubelet-dir
+            - mountPath: /csi
```
Comment (aermakov-zalando) [inline cluster/manifests/03-ebs-csi/02-csi-driver.yaml:76]: `/var/run/csi`? This way it'll be cleaned up on reboots.

File: cluster/manifests/03-ebs-csi/02-csi-driver.yaml [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+            - name: DRIVER_REG_SOCK_PATH
+              value: /opt/podruntime/kubelet/plugins/ebs.csi.aws.com/csi.sock
+          image: registry.opensource.zalan.do/teapot/csi-node-driver-registrar:v1.1.0-master-1
+          lifecycle:
+            preStop:
+              exec:
+                command:
+                  - /bin/sh
+                  - -c
+                  - rm -rf /registration/ebs.csi.aws.com-reg.sock /csi/csi.sock
```
Comment (aermakov-zalando) [inline cluster/manifests/03-ebs-csi/02-csi-driver.yaml]: See above, if we only cleanup in `preStop` it'll be in a broken state after a node reboot.

File: cluster/manifests/03-ebs-csi/ebs-controller.yaml [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+              protocol: TCP
+          resources:
+            requests:
+              cpu: 50m
+              memory: 50Mi
+            limits:
+              cpu: 50m
+              memory: 50Mi
+          volumeMounts:
+            - mountPath: /var/lib/csi/sockets/pluginproxy/
```
Comment (aermakov-zalando) [inline cluster/manifests/03-ebs-csi/ebs-controller.yaml:51]: Same here, `.sock` files are usually in `/var/run`.

File: cluster/cluster.yaml (hunk 1 above)
Comment (arjunrn) [reply cluster/cluster.yaml:1369]: So the worker role is necessary right now because of how `kube-aws-iam-controller` gets credentials right now. It assumes the worker's role first to keep compatibility with kube2iam. So the `WorkerIAMRole` needs to be present for now. Once we move to pod identity credentials then this can be removed.

File: cluster/manifests/03-ebs-csi/02-csi-driver.yaml (hunk 5 above)
Comment (arjunrn) [reply cluster/manifests/03-ebs-csi/02-csi-driver.yaml:2]: I will tie it to the same config item which enables the CSI backed PVCs for prometheus.

File: cluster/manifests/03-ebs-csi/02-csi-driver.yaml (hunk 6 above)
Comment (arjunrn) [reply cluster/manifests/03-ebs-csi/02-csi-driver.yaml:43]: Can be used for when the application cannot accept a secret through a mounted environment variable.

File: cluster/manifests/03-ebs-csi/02-csi-driver.yaml (hunk 8 above)
Comment (arjunrn) [reply cluster/manifests/03-ebs-csi/02-csi-driver.yaml]: What if we change it to a `preStart` hook?

File: cluster/manifests/03-ebs-csi/ebs-controller.yaml (hunk 9 above)
Comment (arjunrn) [reply cluster/manifests/03-ebs-csi/ebs-controller.yaml:51]: This socket is not mounted from the host but is an `emptyDir` so node restarts should not be an issue.

File: cluster/manifests/03-ebs-csi/02-csi-driver.yaml (hunk 8 above)
Comment (aermakov-zalando) [reply cluster/manifests/03-ebs-csi/02-csi-driver.yaml]: That should work. We can also use an initcontainer for this.

Comment (arjunrn) [reply cluster/manifests/03-ebs-csi/02-csi-driver.yaml]: I will change it to an `initContainer`. Also the socket file needs to be that particular directory for kubelet to talk to it.

File: cluster/cluster.yaml (hunk 1 above)
Comment (aermakov-zalando) [reply cluster/cluster.yaml:1369]: I see, thanks!

File: cluster/manifests/03-ebs-csi/02-csi-driver.yaml (hunk 7 above)
Comment (arjunrn) [reply cluster/manifests/03-ebs-csi/02-csi-driver.yaml:76]: This was addressed by having an `initContainer`. See other comment.

**Comment by arjunrn** [comment]
:+1:

File: cluster/cluster.yaml (hunk 1 above)
Comment (linki) [reply cluster/cluster.yaml:1369]: Can you add a note that this can be removed as a comment somewhere (Also fine in the pod identity issue)? Otherwise I think it'll just stick around b/c no one knows what is used for what.

**Comment by jrake-revelant** [comment]
👍
